### PR TITLE
heat_map_analysis update to allow selecting misclassified images and images by id or filename

### PR DESCRIPTION
**Pull request by adstev**
I used some changes from Alex pull request and combined with my own. 

heat_map_analysis has the following additional features:

- set img_type to 'A', 'M' or 'C' to get all, misclassified or correctly classified images respectively
-pass a list of _id_ ' s or a singing _id_ value to image_id to run heatmap on specific images
- pass a list of filenames or a single filename to get heatmaps on the images containing that filename
  image_id is unique while filename is not thus image_id overrides filename if both are passed into the function
- can also display more or less than 5 images by setting max_display

Additionally:
-_id_ was added to the ImageTable
-the model.py attributes documentation was updated
-An attribute valid_res_tbl was added so that the results from predict() can be further analized without bringing to the client
-_id_ column was added to valid_res SASDataFrame output

### CODE REVIEW

**Comment by sasxianhu** [comment]
Kevin, Alex, Wenyu,

Please help review. Thanks.

Best,
XQ

On Jun 1, 2018, at 5:42 PM, adstev <notifications@github.com<mailto:notifications@github.com>> wrote:


EXTERNAL

I used some changes from Alex pull request and combined with my own.

heat_map_analysis has the following additional features:

  *   set img_type to 'A', 'M' or 'C' to get all, misclassified or correctly classified images respectively
-pass a list of id ' s or a singing id value to image_id to run heatmap on specific images
  *   pass a list of filenames or a single filename to get heatmaps on the images containing that filename
image_id is unique while filename is not thus image_id overrides filename if both are passed into the function
  *   can also display more or less than 5 images by setting max_display

Additionally:
-id was added to the ImageTable
-the model.py attributes documentation was updated
-An attribute valid_res_tbl was added so that the results from predict() can be further analized without bringing to the client
-id column was added to valid_res SASDataFrame output

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/sassoftware/python-dlpy/pull/9

Commit Summary

  *   added _id_ column to ImageTable, added _id_ to model.valid_res, added attribute to model of CASTable after running model.predict, changed model.heat_map_analysis to look at misclassified images, images specified by a filename or image_id and changed display to output prediction results
  *   Removed actual filepaths from Example 5 -Heatmaps.ipynb

File Changes

  *   M dlpy/images.py<https://github.com/sassoftware/python-dlpy/pull/9/files#diff-0> (4)
  *   M dlpy/model.py<https://github.com/sassoftware/python-dlpy/pull/9/files#diff-1> (299)
  *   A examples/Example 4 - Heatmaps.ipynb<https://github.com/sassoftware/python-dlpy/pull/9/files#diff-2> (1366)

Patch Links:

  *   https://github.com/sassoftware/python-dlpy/pull/9.patch
  *   https://github.com/sassoftware/python-dlpy/pull/9.diff

**Comment by yg2288** [comment]
Thanks Adam, looks good.

Previously in heat_map_analysis, I could pass gpu= parameter to speed up scoring, but now I'm getting this error:

ERROR: Parameter 'gpu' is not recognized.
ERROR: Expecting one of the following: table, display, outputTables, seed, sampPct, sampPct2, partInd, attributes, diagnostics, output, fixedObs, displayOut, partition, attribute.
ERROR: The action stopped due to errors.

**Comment by adstev** [comment]
Also the is an example notebook which shows the functionality.

**Comment by adstev** [comment]
Hi Alex,

Thanks for letting me know. Could you send me the example? I have never used the gpu functionality and am not sure how it works. I will look into it further.

-Adam

File: dlpy/model.py [hunk 1]
```diff
@@ -1009,8 +1063,55 @@ def heat_map_analysis(self, data, mask_width=None, mask_height=None,
         Returns
         -------
         :class:`pandas.DataFrame`
+            Contains Columns: ['I__label_', 'P__label_Dolphin', 'P__label_Giraffe', '_filename_0',
```
Comment (kesmit13) [inline dlpy/model.py]: I'm assuming not all returned objects have columns 'P__label_Dolphin' / 'P__label_Giraffe'

File: dlpy/model.py [hunk 2]
```diff
@@ -1009,8 +1063,55 @@ def heat_map_analysis(self, data, mask_width=None, mask_height=None,
...
+           '_id_', '_image_', '_label_', 'heat_map']
 
         '''
+
+        # if data is passed we get_predictions if data=None we use model.predict() results
+        def get_predictions(data=data, inputs=inputs, target=target, kwargs=kwargs):
+            input_tbl_opts = input_table_check(data)
+            input_table = self.conn.CASTable(**input_tbl_opts)
+            if target not in input_table.columninfo().ColumnInfo.Column.tolist():
```
Comment (kesmit13) [inline dlpy/model.py]: It's better to use ```input_table.columns``` rather than going through the full action call.

File: dlpy/model.py [hunk 3]
```diff
@@ -1009,8 +1063,55 @@ def heat_map_analysis(self, data, mask_width=None, mask_height=None,
...
+                raise ValueError('Column name "{}" not found in the data table.'.format(target))
+
+            if inputs not in input_table.columninfo().ColumnInfo.Column.tolist():
```
Comment (kesmit13) [inline dlpy/model.py]: Same as above.

**Pull request merged**